dev/Common/AppModel.Identity.h

```diff
@@ -109,6 +112,120 @@ inline winrt::Windows::System::ProcessorArchitecture ParseShortArchitecture(_In_
     }
 }
 
+class PackageVersion : public PACKAGE_VERSION
+{
+public:
+    PackageVersion()
+    {
+        Version = 0;
+    }
+
+    PackageVersion(const PackageVersion&) = default;
+
+    // Create an instance with the value `major.minor.build.revision`.
+    PackageVersion(std::uint16_t major, std::uint16_t minor = 0, std::uint16_t build = 0, std::uint16_t revision = 0) :
+        PACKAGE_VERSION()
+    {
+        Major = major;
+        Minor = minor;
+        Build = build;
+        Revision = revision;
+    }
+
+    // Create an instance from a version as a uint64.
+    PackageVersion(std::uint64_t version)
+    {
+        Version = version;
+    }
+
+// NOTE: There's no preprocessor symbol for Windows.ApplicationModel.PackageVersion
+//       or Windows.ApplicationModel.h so test for Windows.ApplicationModel.IPackageId
+//       as a close proxy for the ABI struct PackageVersion is defined.
+#if defined(____x_ABI_CWindows_CApplicationModel_CIPackageId_INTERFACE_DEFINED__)
+    PackageVersion(ABI::Windows::ApplicationModel::PackageVersion packageVersion) :
+        PACKAGE_VERSION()
+    {
+        Major = packageVersion.Major;
+        Minor = packageVersion.Minor;
+        Build = packageVersion.Build;
+        Revision = packageVersion.Revision;
+    }
+#endif // defined(____x_ABI_CWindows_CApplicationModel_CIPackageId_INTERFACE_DEFINED__)
+
+#if defined(WINRT_Windows_ApplicationModel_2_H)
+    PackageVersion(winrt::Windows::ApplicationModel::PackageVersion packageVersion) :
+        PACKAGE_VERSION()
+    {
+        Major = packageVersion.Major;
+        Minor = packageVersion.Minor;
+        Build = packageVersion.Build;
+        Revision = packageVersion.Revision;
+    }
+#endif // defined(WINRT_Windows_ApplicationModel_2_H)
+
+    PackageVersion& operator=(const PackageVersion&) = default;
+
+    // Return the version as a uint64.
+    std::uint64_t ToVersion() const
+    {
+        return Version;
+    }
+
+#if defined(____x_ABI_CWindows_CApplicationModel_CIPackageId_INTERFACE_DEFINED__)
+    ABI::Windows::ApplicationModel::PackageVersion ToPackageVersion() const
+    {
+        return ABI::Windows::ApplicationModel::PackageVersion{ Major, Minor, Build, Revision };
+    }
+#endif // defined(____x_ABI_CWindows_CApplicationModel_CIPackageId_INTERFACE_DEFINED__)
+
+#if defined(WINRT_Windows_ApplicationModel_2_H)
+    winrt::Windows::ApplicationModel::PackageVersion ToWinrtPackageVersion() const
+    {
+        return winrt::Windows::ApplicationModel::PackageVersion{ Major, Minor, Build, Revision };
+    }
+#endif // defined(WINRT_Windows_ApplicationModel_2_H)
+
+#if defined(_XSTRING_) && defined(_STRSAFE_H_INCLUDED_) && defined(WI_VERIFY)
```

Comment:
Again, isn't this a core piece of the WASDK _implementation_ ? Why use the if guard here? If this is going into the SDK to be used by a C++ consumer, it's also OK to just require "to use PackageVersion you have to include string."  Can you use https://en.cppreference.com/w/cpp/utility/format/format instead so it's `return std::format(...)` ? Be idiomatic when possible.

Reply:
I forgot WinAppSDK switched to C++20. Updating to std::format()

Speaking of...would this work

```
const auto v{ Microsoft::Windows::ApplicationModel::PackageVersion(1,2,3,4) };
const auto message{ std::format(L"This is {}", v) };  //=>"This is 1.2.3.4"
```

if this was defined

```
template <>
struct std::formatter<PACKAGE_VERSION>
{
    constexpr auto parse(std::format_parse_context& ctx)
    {
        return ctx.begin();
    }
    auto format(const PACKAGE_VERSION& version, std::format_context& ctx)
    {
        std::format(std::format_to(ctx.out(), "{}.{}.{}.{}", version.major, version.minor, version.build, version.revision);
    }
};
```

?

Reply:
FYI I just learned about [/Zc:__cplusplus](https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170#remarks)

I also learned [here](https://learn.microsoft.com/cpp/preprocessor/predefined-macros?view=msvc-170)  `_MSVC_LANG` "is set to 202002L if the [/std:c++20](https://learn.microsoft.com/cpp/build/reference/std-specify-language-standard-version?view=msvc-170) compiler option is specified".

Moral of the story, if you need to test the C++ language version use both

```
#if (__cplusplus >= 202002L) || ((defined(_MSVC_LANG) && (_MSVC_LANG >= 202002L)))
```